### PR TITLE
fix(weave): Refactor @op deco to work with `inspect` lib (`isfunction`, `iscoroutinefunction`)

### PR DESCRIPTION
**Pull request by andrewtruong**
### Resolves
- https://wandb.atlassian.net/browse/WB-19428
- https://wandb.atlassian.net/browse/WB-19461
- https://wandb.atlassian.net/browse/WB-19462
- https://wandb.atlassian.net/browse/WB-19686

This PR makes the `@op` deco work with `inspect.isfunction` and `inspect.iscoroutinefunction` which are commonly used in other patching libs

There are a few tradeoffs, discussed here:
https://www.notion.so/wandbai/Weave-op-decorator-changes-a9fcd15d75ec4b1b917bbe657a3b8bd7?pvs=4

### CODE REVIEW

File: weave/flow/agent.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-from typing import Any
```
Comment (shawnlewis) [inline weave/flow/agent.py:1]: Why this change? Maybe just revert this file?

File: weave/legacy/ops_arrow/ops.py [hunk 2]
```diff
@@ -1,11 +1,11 @@
 from weave.legacy.ops_arrow.arraylist_ops import *
 from weave.legacy.ops_arrow.boolean import *
-from weave.legacy.ops_arrow.convert_ops import *
-from weave.legacy.ops_arrow.date import *
+from weave.legacy.ops_arrow.convert_ops import *  # type: ignore
```
Comment (shawnlewis) [inline weave/legacy/ops_arrow/ops.py:3]: unrelated?

File: weave/legacy/ops_primitives/__init__.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-from .. import context_state as _context_state
+from weave.legacy import context_state as _context_state
```
Comment (shawnlewis) [inline weave/legacy/ops_primitives/__init__.py:1]: unrelated?

File: weave/tests/test_op2.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
+import inspect
```
Comment (shawnlewis) [inline weave/tests/test_op2.py:1]: Can this file have a better name?

File: weave/trace/vals.py [hunk 5]
```diff
@@ -145,6 +146,12 @@ def attribute_access_result(
 ) -> Any:
     # Not ideal, what about properties?
     if callable(val_attr_val):
+        sig = inspect.signature(val_attr_val)
```
Comment (shawnlewis) [inline weave/trace/vals.py:149]: I think this code should be a function in op.py called like bind_op()

File: weave/weave_client.py [hunk 6]
```diff
@@ -409,7 +413,16 @@ def create_call(
                 self._anonymous_ops[op] = _build_anonymous_op(op)
             op = self._anonymous_ops[op]
         if isinstance(op, Op):
-            op_def_ref = self._save_op(op)
+            if isinstance(op, partial):
```
Comment (shawnlewis) [inline weave/weave_client.py:416]: can this block be written as a function in op.py?

File: weave/trace/op.py [hunk 7]
```diff
@@ -256,3 +62,264 @@ def _apply_fn_defaults_to_inputs(
...
+    c = _create_call(op, *args, **kwargs)
+    return _execute_call(op, c, *args, **kwargs)
+
+
+def calls(op: Op) -> "CallsIter":
+    client = client_context.weave_client.require_weave_client()
+    return client._op_calls(op)
+
+
+# Legacy decos
```
Comment (shawnlewis) [inline weave/trace/op.py:209]: What is legacy decos? We don't need this. legacy.decorator_op has the legacy decorator, but we shouldn't need to redefine anything here.

File: weave/tests/test_client_trace.py [hunk 8]
```diff
@@ -1316,8 +1316,10 @@ def op_with_custom_type(self, v):
     assert obj2.op_with_custom_type(1) == 2
 
     my_op_ref = weave_client.get_ref(CustomType.op_with_custom_type)
-    with pytest.raises(MissingSelfInstanceError):
-        my_op2 = my_op_ref.get()
+    assert my_op_ref is None, "Must call get on the instance!"
```
Comment (tssweeney) [inline weave/tests/test_client_trace.py:1319]: Love seeing this! Huge (:

File: weave/tests/test_weaveflow.py [hunk 9]
```diff
@@ -256,14 +256,15 @@ def get_item(row):
     assert res == {"in": 1, "out": 1}
 
 
-def test_op_method_name(client):
-    class MyModel(weave.Model):
-        a: int
+class MyModel(weave.Model):
```
Comment (tssweeney) [inline weave/tests/test_weaveflow.py:259]: is this just style or is there an issue with defining the class inside the func?

File: weave/tests/test_op2.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    @op
+    async def amethod(self, a: int) -> int:
+        return a + 1
+
+
+a = A()
+b = B()
+
+
+def test_sync_func(client):
```
Comment (tssweeney) [inline weave/tests/test_op2.py:52]: It would be really nice if all these tests also had a version where the user publishes the object / op first, then "gets" the op and uses it.

File: weave/trace/op.py [hunk 11]
```diff
@@ -256,3 +62,264 @@ def _apply_fn_defaults_to_inputs(
...
+    The interface for Op-ified functions and methods.
+
+    Op was previously a class, and has been converted to a Protocol to allow
+    functions to pass for Op.  This is needed because many popular packages are
+    using the `inspect` module for control flow, and Op instances don't always
+    pass those checks.  In particular, `inspect.iscoroutinefunction` always
+    fails for classes, even ones that implement async methods or protocols.
+
+    Some of the attributes are carry-overs from when Op was a class.  We should
+    consider removing the unnecessary ones where possible.
```
Comment (tssweeney) [inline weave/trace/op.py:79]: it would be nice to note explicitly which of them fall into this category

File: weave/trace/op.py [hunk 12]
```diff
@@ -256,3 +62,264 @@ def _apply_fn_defaults_to_inputs(
...
+    name: str
+    signature: inspect.Signature
+    ref: Optional[ObjectRef]
+    resolve_fn: Callable
+
+    call: Callable[..., Any]
+    calls: Callable[..., "CallsIter"]
+
+    # not sure if this is the best place for this, but kept for compat
+    _set_on_output_handler: Callable[[OnOutputHandlerType], None]
```
Comment (tssweeney) [inline weave/trace/op.py:91]: Both of these are used in integrations - the contract needn't be maintained if you have alternative ways of achieving the same outcome

File: weave/trace/op.py [hunk 13]
```diff
@@ -256,3 +62,264 @@ def _apply_fn_defaults_to_inputs(
...
+    __self__: Any
+
+
+def _set_on_output_handler(func: Op, on_output: OnOutputHandlerType) -> None:
+    if func._on_output_handler is not None:
+        raise ValueError("Cannot set on_output_handler multiple times")
+    func._on_output_handler = on_output
+
+
+def _is_method_alt(func: Callable) -> bool:
```
Comment (tssweeney) [inline weave/trace/op.py:104]: what does this function do?

File: weave/trace/op.py (hunk 7 above)
Comment (tssweeney) [reply weave/trace/op.py:209]: would it be possible to put these in a different file and selectively import??

Comment (andrewtruong) [reply weave/trace/op.py:222]: we can try, but let's do that in a different PR.  This has enough dangerous changes as is!

File: weave/trace/vals.py (hunk 5 above)
Comment (andrewtruong) [reply weave/trace/vals.py]: clarified as `maybe_bind_method` in `op.py`

File: weave/weave_client.py (hunk 6 above)
Comment (andrewtruong) [reply weave/weave_client.py]: clarified as `maybe_unbind_method` in `op.py`

File: weave/trace/op.py (hunk 13 above)
Comment (andrewtruong) [reply weave/trace/op.py]: clarified as `_is_unbound_method` in `op.py`

**Review by tssweeney — APPROVED** [review]
🚢
